### DISCUSSION TITLE
Long running application: Close to tray icon

### DISCUSSION THREAD

**Discussion by jdheeter** [tauri-apps/tauri#2684] [Q&A] [answered]
Many applications need to be able to operate in the background without a visible window. A good example of this would be an app like https://justgetflux.com

Currently we have implemented a tray icon, but on windows when the main windows is closed (clicking the window X) the application is killed. On MacOS, when the main window is closed the application will continue to be run (can be closed using the tray icon menu). There is an event you can listen to inside Tauri called CloseRequested, but I don't see any way to disable the logic that closes the entire Tauri application when the window is closed. 

Is there any good reference/example for a Tauri application that closes to the tray?

**Comment by amrbashir** [comment, marked as answer]
Check this https://github.com/tauri-apps/tauri/blob/51dd08ab254461000867ce4ad4d29ce2ecc6194c/examples/api/src-tauri/src/main.rs#L233

**Reply by lucasfernog** [reply]
The right link is https://github.com/tauri-apps/tauri/blob/51dd08ab254461000867ce4ad4d29ce2ecc6194c/examples/api/src-tauri/src/main.rs#L233

**Reply by jdheeter** [reply]
Ok, I have followed the example and it works great. So the window can be hidden instead of closed, and the menu label can be updated from "Hide" to "Show" to properly reflect the status of the main window. 

The only issue I'm seeing at this point is that for applications with no active window it's normal that the application should only be visible in the tray (the icon should no longer be in the main taskbar unless a window is open.) Is there a Tauri Rust API to remove the application icon from the OS taskbar so the app is only visible in the tray? Otherwise, it might be good to have the application icon visible in the taskbar + tray (maybe the user could configure it) but in this circumstance, it would be good to .show() the window when the user clicks on the taskbar icon. 

Currently when the window is hidden the only way to see it is to use the tray menu "show" selection. How do I trigger .show() when the user clicks on the icon in the main taskbar? Currently clicking on the icon does nothing.

I tried activation policy, but this causes the main taskbar icon to always be hidden (it should be visible when a window is visible). Also I haven't tested to see how this performs on another OS.
``` 
#[cfg(target_os = "macos")]
app.set_activation_policy(tauri::ActivationPolicy::Accessory);
```

Tested on Windows and the taskbar icon acts as expected (menu icon disappears when window is hidden, tray icon remains visible as long as app is still running).

**Reply by joepio** [reply]
I think the original answer may be outdated. `Event` has been split into various types, and `RunEvent::ExitRequest` has been removed in rc5.

Not sure if there is an API right now to keep the app open after closing the last window?

**Reply by amrbashir** [reply]
@joepio `RunEvent::ExitRequested` stilll exists, `RunEvent::ClosRequested` has been moved to `RunEvent::WindowEvent::CloseRequested`.
You can find all the events here https://docs.rs/tauri/1.0.0-rc.6/tauri/enum.RunEvent.html

**Reply by pacholoamit** [reply]
As of tauri v1.1. Here's what I came up with

```rust

tauri::Builder::default().on_window_event(|event| match event.event() {
            tauri::WindowEvent::CloseRequested { api, .. } => {
                event.window().hide().unwrap();
                api.prevent_close();
            }
            _ => {}
        }) //additional code for tauri

```

**Reply by rizaldirnm** [reply]
How do we listen when user want to show again? @pacholoamit

**Reply by yuhr** [reply]
@rizaldirnm [This article](https://betterprogramming.pub/create-menubar-app-with-tauri-510ab7f7c43d#ec7e) may help. Here's a slightly modified version (not to use `tauri-plugin-positioner`) of the example code from the section:

```rust
		.on_system_tray_event(|app, event| match event {
			SystemTrayEvent::LeftClick { position: _, size: _, .. } => {
				let window = app.get_window("main").unwrap();
				if window.is_visible().unwrap() {
					window.hide().unwrap();
				} else {
					window.show().unwrap();
					window.set_focus().unwrap();
				}
			}
		})
```

**Reply by benediktms** [reply]
Is it possible to handle intercepting the event when a user quits the application via OS shortcuts (Cmd+Q/Alt+F4)? My assumption is that this gets handled by the OS so I'd have to hook into those APIs myself, is that correct? Or does Tauri provide a way to do this?

**Comment by dQw4w9WgXcQ** [comment]
`app.set_activation_policy(ActivationPolicy::Accessory);`

**Comment by davidklein147** [comment]
I also faced this problem when I migrated my app from v1 to v2

I found that when the "X" of the window is clicked then "tauri::RunEvent::ExitRequested with code = None" triggered, so when I click exit on the tray icon I give it code 0, then I handle closing the app according to the code, when the window closed anyway

Here is my solution:

```
fn tray_event_handler(app: &AppHandle, event: MenuEvent, url: Arc<Url>) {
    match event.id.as_ref() {
        "open" => {
            let window = app.get_webview_window("main");
            match window {
                Some(w) => {
                    let win_is_open = w.is_visible().unwrap();
                    match win_is_open {
                        false => {
                            info!("show window");
                            w.show().unwrap();
                        }
                        true => {}
                    }
                }
                None => {
                    let _ = build_main_window(app, url);
                    ()
                }
            }
        }
        "quit" => {
            let app = app.app_handle();
            app.exit(0)
        }
        _ => (),
    }
}
```

```
fn run_event_handler(_app_handle: &AppHandle, event: tauri::RunEvent) {
    match event {
        tauri::RunEvent::ExitRequested { code, api, .. } =>
            match code {
                Some(c) => {
                    info!("Exit Request with code: {}", c);
                }
                _ => {
                    info!("Exit Request with code: {:?} - prevented exit", code);
                    api.prevent_exit();
                }
            }
        _ => (),
    }
}
```

**Comment by data-miner00** [comment]
For the newer version of Tauri (as of 2025), the updated code are as below

```rs
tauri::Builder::default()
    .on_window_event(|window, event| {
        match event {
            tauri::WindowEvent::CloseRequested { api, .. } => {
                window.hide().unwrap();
                api.prevent_close();
            }
            _ => {}
        }
    })
```
